sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RedirectPolicy.java

```diff
@@ -4,76 +4,210 @@
+        if ((shouldRedirectCondition != null && shouldRedirectCondition.test(requestRedirectCondition))
+            || (shouldRedirectCondition == null && defaultShouldAttemptRedirect(requestRedirectCondition))) {
+            createRedirectRequest(response);
+            return attemptRedirect(next, redirectAttempt + 1, attemptedRedirectUrls);
+        }
+
+        return response;
+    }
+
+    public boolean defaultShouldAttemptRedirect(RequestRedirectCondition requestRedirectCondition) {
+        Response<?> response = requestRedirectCondition.getResponse();
+        int tryCount = requestRedirectCondition.getTryCount();
+        Set<String> attemptedRedirectUrls = requestRedirectCondition.getRedirectedUrls();
+        String redirectUrl = response.getHeaders().getValue(this.locationHeader);
+
+        if (isValidRedirectStatusCode(response.getStatusCode())
+            && isValidRedirectCount(tryCount)
+            && isAllowedRedirectMethod(response.getRequest().getHttpMethod())
+            && redirectUrl != null
+            && !alreadyAttemptedRedirectUrl(redirectUrl, attemptedRedirectUrls)) {
+
+            LOGGER.atVerbose()
+                .addKeyValue(LoggingKeys.TRY_COUNT_KEY, tryCount)
+                .addKeyValue(REDIRECT_URLS_KEY, attemptedRedirectUrls::toString)
+                .log(() -> "Redirecting.");
+
+            attemptedRedirectUrls.add(redirectUrl);
+
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Check if the attempt count of the redirect is less than the {@code maxAttempts}
+     *
+     * @param tryCount the try count for the HTTP request associated to the HTTP response.
+     *
+     * @return {@code true} if the {@code tryCount} is greater than the {@code maxAttempts}, {@code false} otherwise.
+     */
+    private boolean isValidRedirectCount(int tryCount) {
+        if (tryCount >= this.maxAttempts) {
+            LOGGER.atError()
+                .addKeyValue("maxAttempts", this.maxAttempts)
+                .log(() -> "Redirect attempts have been exhausted.");
+
+            return false;
+        }
+
+        return true;
+    }
 
-            return attemptRedirect(redirectRequestCopy, next, redirectAttempt + 1, attemptedRedirectUrls);
+    /**
+     * Check if the redirect url provided in the response headers is already attempted.
+     *
+     * @param redirectUrl the redirect url provided in the response header.
+     * @param attemptedRedirectUrls the set containing a list of attempted redirect locations.
+     *
+     * @return {@code true} if the redirectUrl provided in the response header is already being attempted for redirect,
+     * {@code false} otherwise.
+     */
+    private boolean alreadyAttemptedRedirectUrl(String redirectUrl,
+                                                Set<String> attemptedRedirectUrls) {
+        if (attemptedRedirectUrls.contains(redirectUrl)) {
+            LOGGER.atError()
```

Comment:
These are the same log levels in azure-core, do we want to change that?
It should be a good practice to avoid redirect loops where a URL redirects to another URL that has already been redirected to. So error it out?